hadoop-hdfs-project/hadoop-hdfs-client/src/main/java/org/apache/hadoop/hdfs/server/namenode/ha/ObserverReadProxyProvider.java

```diff
@@ -342,7 +342,7 @@ private synchronized void initializeMsync() throws IOException {
     if (msynced) {
       return; // No need for an msync
     }
-    getProxyAsClientProtocol(failoverProxy.getProxy().proxy).msync();
```

Comment:
what does failoverProxy.getProxy().proxy points to?
the same logic seems to be used for invocation as well
```
      // Either all observers have failed, observer reads are disabled,
      // or this is a write request. In any case, forward the request to
      // the active NameNode.
      LOG.debug("Using failoverProxy to service {}", method.getName());
      ProxyInfo<T> activeProxy = failoverProxy.getProxy();
      try {
        retVal = method.invoke(activeProxy.proxy, args);
      } catch (InvocationTargetException e) {
        // This exception will be handled by higher layers
        throw e.getCause();
      }
```

Reply:
hi， @ayushtkn , thanks a lot for replying.
In class ObserverReadProxyProvider, there are two varaibles used to get NameNode proxy. One is failoverProxy, another is nameNodeProxies. For read requests,  it will use nameNodeProxies to get a proxy of observer namenode, but for msync rpc, it will use varaible failoverProxy to get an active namenode proxy.

Reply:
hi， @ayushtkn , thanks a lot for replying.
In class ObserverReadProxyProvider, there are two varaibles used to get NameNode proxy. One is failoverProxy, another is nameNodeProxies. For read requests, it will use nameNodeProxies to get a proxy of observer namenode, but for msync rpc, it will use varaible failoverProxy to get an active namenode proxy.

Reply:
hi， @ayushtkn , thanks a lot for replying.
In class ObserverReadProxyProvider, there are two varaibles used to get NameNode proxy. One is failoverProxy, another is nameNodeProxies. For read requests, it will use nameNodeProxies to get a proxy of observer namenode, but for msync rpc, it will use varaible failoverProxy to get an active namenode proxy.

In practice, we found a lots of exceptions in  the observer namenode log like following:

`org.apache.hadoop.hdfs.protocol.ClientProtocol.msync from 10.152.115.166:59292
org.apache.hadoop.ipc.ObserverRetryOnActiveException: Operation category WRITE is not supported in state observer. Visit https://s.apache.org/sbnn-error
`
we found that failoverProxy contains proxys of all kinds of namenode, so we want to find active namenode proxy before invoke msync rpc. There has no effect on sending read requests to observer namenode, because read request use nameNodeProxies find observer namenode rather than failoverProxy.

Reply:
> 
@ayushtkn 
failoverProxy point to a random namenode of all.  But in fact, we don't want it to point to observer namenode.

Reply:
@xkrogen , hi, erik, could you please help me look at this ?